Migrate existing Azure AI Face service to TypeSpec

Han-msft opened the pull request:
# Data Plane API - Pull Request
Face API is an existing service & restful API for 5+ years as documented in [What is the Azure AI Face service?](https://learn.microsoft.com/en-us/azure/ai-services/computer-vision/overview-identity). We continue to extend with new API and the latest addition was [Detect liveness in faces](https://learn.microsoft.com/en-us/azure/ai-services/computer-vision/tutorials/liveness) that was released in Ignite in Nov 2023 as a Public Preview. At that time, we wanted to release a REST SDK for detect liveness too. However, we were told we cannot use OpenAPI spec that we used to create SDK for old APIs like Detect/Verify/Identify, etc. anymore and TypeSpec is the only option to release new SDK. Our current objective is to release an All-In-One SDK for all the APIs by using TypeSpec, including the existing Face Detect, Verify, Identify, FindSimilar, and the latest detect liveness.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/Face/preview/v1.0-preview/Face.json
* Updated paths: https://github.com/Han-msft/azure-rest-api-specs/blob/hachiang/face-typespec/specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

Commit:
Add following operations:
- Liveness Session
- Person Directory
- Person Group
- Face List
- Face Operations

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+        "responses": {
+          "200": {
+            "description": "The request has succeeded.",
+            "schema": {
+              "type": "array",
+              "items": {
+                "$ref": "#/definitions/LivenessSessionItem"
+              }
+            }
```
Response schema should not be a bare array.

Han-msft replied:
This API is already public previewed. We prefer not to change it.

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+    "/face/{apiVersion}/detectLiveness/singleModal/sessions": {
+      "get": {
+        "operationId": "LivenessSessions_ListSessions",
+        "description": "Operation that lists resources in a paginated way.",
```
How does pagination work for this API?

Han-msft replied:
It's the description that comes from the TypeSpec Azure.Core operation. Our list operation only support query by top and start, I think it's not fulfilling the pagination of Azure API. I'll override the description.

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+    "ListRequestOptions.top": {
+      "name": "top",
+      "in": "query",
+      "description": "The number of sessions to list, ranging in [1, 1000]. Default is 1000.",
```
Top should not have a default value. The default is to return all elements of the collection, paginated as needed.

Han-msft replied:
Our long time ago GAed API use these parameters for list request. We'll consider revise it in our next generation service but not changing it for now.

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+    "ListRequestOptions.start": {
+      "name": "start",
+      "in": "query",
+      "description": "List sessions from the last sessionId greater than the \"start\". It contains no more than 64 characters. Default is empty.",
```
This suggests that sessions are returned sorted by sessionId -- is that correct? If so that should be documented.

Han-msft replied:
I moved the description to `BaseSessions.listSessions` since ListRequestOptions is shared by other GAed operations, it's not limited to session.

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+  },
+  "tags": [],
+  "paths": {
+    "/face/{apiVersion}/compare": {
```
I think you should move "/face/{apiVersion}" into the `x-ms-parameterized-host`, as it was in your GA API. That will be more DRY and also make comparison across API versions easier.

Han-msft replied:
Updated. However, `Azure.Core.Foundations.ApiVersionParameter` is injected to all our API's parameter, I could not find a way to remove it.

Han-msft replied:
I'm going to rollback the change since there is no way to remove ApiVersionParameter in query if we don't override it by putting it in request path.

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+    }
+  ],
+  "securityDefinitions": {
+    "ApiKeyAuth": {
```
Please include a description in each of these securityDefinition for the generated REST API reference docs. 

See the [TypeSpec docs] for an example of using the @doc decorator to specify a description for the securityDefinition. 

[TypeSpec docs]: https://azure.github.io/typespec-azure/docs/reference/azure-style-guide#examples

Han-msft replied:
Updated.

mikekistler commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
```diff
@@ -0,0 +1,7948 @@
+{
```
I think you should populate the GA API in the "stable/v1.0" folder, to establish the baseline for tooling.

There are two ways to do this. One is to make your TypeSpec versioned, with a v1.0 version and a v1.1-preview version. I think this is the preferred solution. Or you could just copy the directory over from the cognitiveservices directory as is.

Han-msft replied:
I thought we only need to support latest version when migrating to TypeSpec. That's what I observed from document intelligence and image analysis. I'll go for copy the swagger file. Please let me know if you think we should include them in TypeSpec.

Han-msft replied:
I copied the stable swagger to current ai folder. How do I know whether it works for tooling? Would it take effect on APIView or it's other tool?

Han-msft replied:
I'm going to revert the swagger commit for now. Because Jeffery said he is going to check more detail about folder change of cognitive service. Copying swagger also bring more validation failure.

mikekistler reviewed (changes requested):
Found a few issues I hope you can address.

heaths commented on specification/ai/Face/models.session.tsp:
```diff
@@ -0,0 +1,187 @@
+
+  #suppress "@azure-tools/typespec-azure-core/casing-style"
+  @doc("The CreationContent for creating session.")
+  CreationContent: LivenessSessionCreationContentForMultipart;
+
+  #suppress "@azure-tools/typespec-azure-core/casing-style"
+  @doc("The image stream for verify.")
+  VerifyImage: bytes;
```
Why did you switch to PascalCase for these two properties?

Han-msft replied:
That's reflecting the underlying key inside the form-data request of our existing service. Do you prefer to name these properties with camelCase and use decorator to specify it's PascalCase in actual request?

heaths commented on specification/ai/Face/models.common.tsp:
```diff
@@ -0,0 +1,196 @@
+namespace Face;
+
+@doc("The request for list resources.")
+model ListRequestOptions {
+  @doc("List resources greater than the \"start\". It contains no more than 64 characters. Default is empty.")
+  @query
+  start?: string;
+
+  @doc("The number of items to list, ranging in [1, 1000]. Default is 1000.")
+  @query
+  top?: int32;
+}
```
These should be query parameters named ~~start~~ `skip` and `top`. That's the recommendation, anyway.

Han-msft replied:
This would be breaking change to existing service. We should consider in future release.

heaths commented on specification/ai/Face/models.common.tsp:
```diff
@@ -0,0 +1,196 @@
+}
+
+@doc("Type of training status.")
+enum TrainingStatusType {
```
We already have a standard enum for LRO status. See https://azure.github.io/typespec-azure/docs/getstarted/azure-core/step06 for an example. The values you define here aren't correct, for example, and should be PascalCase.

Han-msft replied:
This is also for operation GAed a long time ago.

heaths commented on specification/ai/Face/models.common.tsp:
```diff
@@ -0,0 +1,196 @@
+
+@doc("Response body for find similar face operation.")
+model FindSimilarResult {
+  @doc("Confidence value of the candidate.")
```
As with other cognitive services, I recommend documenting this is between 0.0 and 1.0.

heaths commented on specification/ai/Face/routes.common.tsp:
```diff
@@ -0,0 +1,278 @@
+
+namespace Face;
+
+alias ServiceTraits = NoClientRequestId &
```
Why do you say you don't support a client request ID here but then have a property for it in your audit models? We do recommend supporting client request IDs to assist callers with tracing, etc.

Han-msft replied:
We support it but not with header key "x-ms-client-request-id", which is the behavior of SupportsClientRequestId trait.

heaths reviewed (approved):
Approved based on the changes we discussed in the meeting, but for a v2 there's a lot to consider here. More than I've reviewed since this PR was not the one listed in the meeting request for review (and I reviewed that one, which was significantly smaller).

Han-msft commented on specification/ai/Face/models.session.tsp:
```diff
@@ -0,0 +1,187 @@
+}
+
+@doc("Session status.")
+enum SessionStatus {
```
No, it's not the status of operation. It stands for the lifecycle of created session object.

Commit:
Restructure and rename operations
Omit unreachable types

Commit:
Update example, doc, data decorator and unnecessary sharedRoute

Commit:
Change PD to non-LRO

weidongxu-microsoft commented on specification/ai/data-plane/Face/preview/v1.1-preview.1/Face.json:
If the new preview is moved here, it seems hard to know if the API is compatible with stable v1.0
https://github.com/Azure/azure-rest-api-specs/tree/f05d05a88ab9f6a58a49859cf753a4c36de6f137/specification/cognitiveservices/data-plane/Face/stable/v1.0

weidongxu-microsoft commented on specification/ai/Face/models.common.tsp:
```diff
@@ -0,0 +1,218 @@
+}
+
+@doc("The recognition model for the face.")
+enum RecognitionModel {
```
extensible enum now need to be written as union

context https://github.com/Azure/azure-rest-api-specs/pull/27687

Han-msft replied:
Fixed

weidongxu-microsoft commented on specification/ai/Face/models.common.tsp:
```diff
@@ -0,0 +1,218 @@
+  persistedFaceId: string;
+
+  @doc("User-provided data attached to the face. The size limit is 1KB.")
+  userData?: string;
```
Is it `bytes` or `string`?

`bytes` in JSON is binary base64 encoded to string.

Han-msft replied:
It's string, corrected the description and added decorator.

weidongxu-microsoft commented on specification/ai/Face/models.common.tsp:
```diff
@@ -0,0 +1,218 @@
+
+  @query
+  @doc("The 'detectionModel' associated with the detected faceIds. Supported 'detectionModel' values include 'detection_01', 'detection_02' and 'detection_03'. The default value is 'detection_01'.")
+  detectionModel?: DetectionModel;
```
server side default could be written as `detectionModel?: DetectionModel = DetectionModel.detection_01;`

Han-msft replied:
Fixed, also applied to most of request parameter.

weidongxu-microsoft commented on specification/ai/Face/routes.persondirectory.tsp:
```diff
@@ -0,0 +1,254 @@
+  @doc("""
+  Retrieve a person's persistedFaceIds representing the registered person face feature(s).
+  """)
+  @sharedRoute
```
nit, there is `@sharedRoute`, but it is pretty hard for me to find what is the other operation that this op shares the endpoint.

Han-msft replied:
Removed, should be some workaround for previous version compiler, it was complaining for same path but different HTTP method operations.

weidongxu-microsoft reviewed (approved):
Mostly LGTM

I assume the folder structure is under discussion.

Also please take a look on LintDiff and ModelValidation (some of them could be invalid), see if any could be a quick fix.

Commit:
Use union to represent extensible enum
Add suppressions
Add default value for models
Fix age type
Rename model in sdk

Refine examples and train headers

Add uuid type, groupId type and name min length

Use our own operation state

Change liveness operation to enum

Refine group id pattern

Use array for csv fields

Separate request and response model for collection

Flatten interface

Flatten parameter for more interface

Han-msft commented:
@AzureRestAPISpecReview
We got `Breaking Change` check failure after we copied our previous version swagger by #28424.
I briefly go through them and believe most of them are annotation difference caused by migrating to TypeSpec.
I also added some suppressions for `Swagger LintDiff`.
Please help take a look.
We'd like to merge it if there is no concern.

Commit:
Add liveness audit response schema

Commit:
Mark detect as internal for customization

shaoli-msft committed:
Update documents (add and fix description format and add examples)

Co-authored-by: Shawn Lin <Shao.Lin@microsoft.com>

Commit:
Move path to host template

Commit:
Move suppression to config

Fix format and suppress response code failure

Fix breaking changes

Switch endpoint to uri

Refine liveness session model

Move away from standard operation to use our own error code

Commit:
Set liveness with verify session as internal

MaryGao commented:
/azp run

shaoli-msft committed:
Remove 4xx examples for now with some document updates (#4)

* Update detection document

* Update 4xx examples

* Update example parameter and response

* Several doc fix for PD

* Differenct recognition descriptions for different data structure

* Fix bullet points in note and update some PD doc

* Update liveness doc

* Remove 4xx examples for now to pass validation

---------

Co-authored-by: Shawn Lin <Shao.Lin@microsoft.com>
Co-authored-by: Shawn Lin <shaoli@microsoft.com>

Commit:
Add verify image response in create session

shaoli-msft committed:
Rename API operation name (#5)

* Change document to new operation names

* Rename

* More renaming

---------

Co-authored-by: Shawn Lin <Shao.Lin@microsoft.com>
Co-authored-by: Shawn Lin <shaoli@microsoft.com>

Commit:
Refine verify image header description

Commit:
Add partial update for java

mikekistler reviewed (approved):
Looks good. 👍

Commit:
Use shared route for DPG LRO
Use original LRO for PD person and face creation

Commit:
Use resource decorator instead of specify resource in LRO response

Commit:
Change name for patch to optional

Commit:
Correct operation status name
Correct DPG example

Commit:
Rename operation status model

Commit:
Correct operation result example

Commit:
Add gated attribute

Commit:
Fix validation

Commit:
Force public model for internal detect

Commit:
Correct update DPG model

Commit:
Disable detect from url convenientAPI for java

Commit:
Rename dpg write

Commit:
Fix comment

Commit:
Rename id and remove customized uuid

Commit:
Correct access level of landmark and attribute

Commit:
Correct more models' access

Commit:
Collect client customization into client.tsp

Commit:
Correct verify pd parameter type

Commit:
Correct type

Commit:
Reorder detect parameters

Commit:
Rename session model

Commit:
Refine description for return reco model

Commit:
Remove detection convenient API

shaoli-msft committed:
Add link of error code and message (#6)

Co-authored-by: Shawn Lin <Shao.Lin@microsoft.com>

Commit:
Fix validation

Commit:
Merge branch 'hachiang/face-typespec' of github.com:Han-msft/azure-rest-api-specs into hachiang/face-typespec

MaryGao commented:
/azp run

Commit:
Remove unnecessary model

MaryGao commented on specification/ai/Face/routes.session.tsp:
```diff
@@ -0,0 +1,152 @@
+  @returnsDoc(SessionCreationSuccess)
+  createLivenessSession is FaceResourceCreateWithServiceProvidedName<
+    LivenessSession,
+    CreateLivenessSessionParameters,
```
Not sure we could rename this model name `CreateLivenessSessionParameters`.

 This is conflicted with RLC's naming rule `{Operations}Parameters`. We would resolve this conflict in emitter side. To non-block you it would be great if you could rename it.

Commit:
Rename model

Commit:
Remove redundant line

Han-msft committed:
Merge branch 'main' into hachiang/face-typespec
